[Poker game step4] 포커 딜러와 게임 테스트

1Consumption opened the pull request:
고민한 점
* Player와 Dealer의 관계
딜러는 카드를 나눠주는 역할도 하지만, 동시에 플레이어의 역할도 합니다. 초기에는 플레이어와 딜러를 서로 독립된 객체로 생각하여 관계를 설정하지 않았지만, 두 클래스가 동일한 역할을 하는 부분이 있어서 Player 클래스를 Dealer 클래스가 상속받게 하였습니다. 이로 인해 중복되는 코드를 줄일 수 있었습니다.
* private한 프로퍼티는 어떻게 테스트할 것인가?
테스트 코드에서 특정 객체의 private한 프로퍼티는 접근이 불가능하여 해당 프로퍼티에 대해 어떻게 접근할지에 대해 고민 했습니다. 이를 위해 private 프로퍼티의 값만 받아오는 인터페이스 메소드를 public으로 선언하였고, 변한 값에 대해 테스트할 수 있었습니다.

1Consumption committed:
model 폴더 추가

card, cardDeck등 model을 담당하는 클래스를 모아두기 위해 폴더 생성 및 클래스 이동

1Consumption committed:
dealer 클래스 추가

player에게 카드를 분배 및 게임을 진행 할 dealer 클래스 생성

1Consumption committed:
dealer 클래스 프로퍼티 및 메소드 추가

전체 카드 덱을 저장하는 cardDeck 프로퍼티 추가
딜러 자신의 카드 덱을 저장하는 ownCardDeck 프로퍼티 추가
player에게 한장의 카드를 나눠주는 deal 메소드 추가

1Consumption committed:
PokerGame class 추가

포커게임에 대한 데이터를 저장하고 로직을 처리 해주기위한 클래스

1Consumption committed:
PokerGame class에 enum 추가

스터드 방식을 제시하는 Stud와 플레이어의 수를 제시하는 NumOfPlayer enum

1Consumption committed:
Player 클래스 추가

Poker game을 플레이할 Player 클래스

1Consumption committed:
Player 클래스 프로퍼티 및 메소드 추가

player의 card deck을 저장 할 ownCardDeck 프로퍼티
카드가 인자로 전해지면 ownCardDeck에 append 하는 receiveCard 메소드
player 인스턴스를 문자열로 반환하기위해 CustomStringConvertible 프로토콜 채택

1Consumption committed:
PokerGame, Dealer클래스 변경
PokerGame
Player 배열 추가
stud의 초기값 7, numOfPlayer의 초기값 4로 지정

Dealer
생성자에서 stud,numOfPlayer값을 받도록 수정

1Consumption committed:
PokerGame, Dealer클래스 변경
PokerGame
gameStart 메소드 추가 : dealer와 players에 카드객체를 전달함
dealCardToPlayer 메소드 추가 : dealer에서 카드 인스턴스를 받아와서 nil인지 체크하고 nil이 아니라면 player에 추가해주고, nil이라면 종료

Dealer
Player 객체를 상속받도록 수정 : ownCardDeck과 receiveCard가 중복됨을 확인하고 변경.

1Consumption committed:
PokerGame, Dealer클래스 변경
PokerGame
readyPlayer 메소드 추가 : stud이나 numOfPlayer가 바뀔 때 player 정보를 변경하는데, 코드 중복을 피하기 위해 생성. player의 정보를 stud,numOfPlayer에 맞게 설정해줌.
setStud, setNumOfPlayer 메소드 추가 : 각각 stud, numOfPlayer의 값을 변경하고, 그에 맞게 player의 정보 갱신

Dealer
stud, numOfPlayer 타입을 enum으로 변경 : 보다 정확히 명시하고 값을 제한하기 위해.

1Consumption committed:
Player 클래스에 test를 위한 extension 추가

private한 ownCardDeck을 테스트하기 위해 테스트 코드에서 사용 가능한 메소드 추가
test case 작성

1Consumption committed:
PokerGame 클래스에 test를 위한 extension 추가
private한 프로퍼티들을 테스트하기 위해 인터페이스 메소드를 추가해줌
test case 작성

1Consumption committed:
readme 수정

추가한 기능 정리

godrm commented on CardGameApp/CardGameApp/model/Dealer.swift:
```diff
@@ -0,0 +1,26 @@
+
+import Foundation
+
+class Dealer:Player{
```
이런 경우 클래스 상속보다는 Dealer와 Player가 같은 프로토콜을 채택하는 것을 권장합니다. 
왜 프로토콜을 채택하는 걸 권장하는 지 비교해보세요.

godrm commented on CardGameApp/CardGameApp/model/Dealer.swift:
```diff
@@ -0,0 +1,26 @@
+        self.cardDeck=CardDeck()
+        self.cardDeck.shuffle()
+        self.stud=stud
+        self.numOfPlayer=numOfPlayer
```
선언할 때도, `:`, `,`, `=` 같은 기호를 띄어쓰기 하는게 어떨까요? 인지하기 어렵네요.

godrm commented on CardGameApp/CardGameApp/model/Dealer.swift:
```diff
@@ -0,0 +1,26 @@
+    private var stud:PokerGame.Stud
+    private var numOfPlayer:PokerGame.NumOfPlayer
+    
+    init(stud:PokerGame.Stud,numOfPlayer:PokerGame.NumOfPlayer){
```
딜러가 이 값을 모두 알고 있는게 좋을까요? dealer보다 상위 객체에서만 알고 있어도 되는건 아닐까요?

godrm commented on CardGameApp/CardGameApp/model/PokerGame.swift:
```diff
@@ -0,0 +1,102 @@
+    private func readyPlayer(){
+        self.dealer=Dealer(stud: self.stud, numOfPlayer: self.numOfPlayer)
+        self.players=[Player]()
+        for _ in 0..<self.numOfPlayer.rawValue{
```
rawValue를 접근해서 값을 사용하기 보다는 NumOfPlayer enum 타입을 그대로 활용하는 게 더 좋습니다. 
enum, struct, class 모두 내부 값을 가져와서 다른 곳에서 활용하거나 비교, 탐색하는 경우가 있습니다.
그런 코드가 타입 내부에 구현되어 있고 호출하는 곳에는 간략하게 보이도록 개선해보세요.
```
enum NumOfPlayer: Int {
case one=1,two,three,four

func forEach(_ transform : () -> ()) {
    for _ in 0..<self.rawValue {
        transform()
    }
}

}
```
이렇게 선언하면 호출할 때는 `numOfPlayer.forEach { players.append(Player()) }` 형태가 됩니다.  
참고하세요.

godrm commented on CardGameApp/CardGameApp/model/PokerGame.swift:
```diff
@@ -0,0 +1,102 @@
+        }
+    }
+    
+    private func dealCardToPlayer(player:Player){
```
프로토콜을 타입으로 활용하면 이 함수 매개변수도 프로토콜을 타입으로 활용할 수 있습니다.

godrm commented on CardGameApp/CardGameApp/model/PokerGame.swift:
```diff
@@ -0,0 +1,102 @@
+        if let card = myCard{
+            player.receiveCard(card: card)
+        } else {
+            exit(0)
```
프로그램 로직 어딘가에서 exit()을 호출하는 것은 지양하세요. 프로그램을 멈추는 것보다는 유지하면서 사용자에게 인지시켜 주는게 중요합니다.

godrm commented on CardGameApp/CardGameApp/model/PokerGame.swift:
```diff
@@ -0,0 +1,102 @@
+    func setNumOfPlayer(numOfPlayer:NumOfPlayer){
+        self.numOfPlayer=numOfPlayer
+        readyPlayer()
+        gameStart()
```
위에 setStud와 코드가 중복되는 것 같은데 값이 바뀌면 이렇게 각각 호출하는 게 좋을까요?

godrm commented on CardGameApp/CardGameApp/model/PokerGame.swift:
```diff
@@ -0,0 +1,102 @@
+        }
+        result+="딜러 \(dealer)\n"
+        return result
+    }
```
게임에 대한 출력을 위해서 CustomStringConvertible 프로토콜로 문자열로 만드는 것은 좋은 선택이 아닙니다.
더 다양한 형태를 지원할 수 있도록 메소드를 만들어보세요.

godrm commented on CardGameApp/CardGameApp/model/PokerGame.swift:
```diff
@@ -0,0 +1,102 @@
+        return self.numOfPlayer
+    }
+}
+#endif
```
이 메소드들은 왜 있는거죠? 아무리 디버거 용이라도 불필요해 보이는 데요? 
테스트를 위해서 만든거면 테스트를 위해서 이렇게 별도로 만들 필요는 없습니다.
실제로 사용하는 메소드를 이용해서 테스트해야 합니다.

godrm reviewed (commented):
다음 피드백을 확인해보세요.

1Consumption committed:
띄어쓰기 적용

선언, 인자값 넘길 때 등 띄어쓰기 적용

1Consumption committed:
Dealer class 불필요 코드 재거

Dealer에서 stud,numOfPlayer가 필요없기 때문에 해당 프로퍼티 제거

1Consumption committed:
MustHaveCard 프로토콜 추가

Player와 Dealer가 카드를 가져야한다는 공통점을 프로토콜로 강제함

1Consumption committed:
Player, Dealer에 MustHaveCard 프로토콜 채용

card에 대한 프로퍼티를 read only로 설정하여 외부에서 값 변경 못하도록 설정
receiveCard() 메소드를 강제하여 기능 구현하도록 유도

1Consumption committed:
NumOfPlayer enum, Stud enum 변경

직접 rawValue에 접근하지 않고 원하는 작업을 할 수 있도록 forEach문 추가하여 원하는 작업을 원하는 rawValue만큼 반복하도록 수정

1Consumption committed:
PokerGame 클래스 변경

카드를 나눠주던 중 카드가 없으면 exit() 을 호출하지 않고 없다고 콘솔창에 출력

1Consumption committed:
PokerGame 클래스 setGameStyle 메소드 추가

기존 stud과 numOfPlayer는 각자 따로 관리하던 코드가 있었으나, 중복되는 점이 많아 두 프로퍼티를 setGameStyle이라는 함수에서 한번에 관리 하도록 변경

1Consumption committed:
PokerGame 클래스 CustomStringConvertibel 프로토콜 취소

대신 printPlayersCard() 메소드에서 String 객체를 반환하도록 수정

1Consumption committed:
사용하지 않는 메소드 제거

1Consumption committed:
Merge branch 'pokerGame-step4' of https://github.com/1Consumption/swift-pokergameapp into pokerGame-step4

1Consumption committed:
PokerGame 클래스에 private 프로퍼티에 접근하기 위한 인터페이스 메소드 추가

dealer와 players 인스턴스의 값을 받아오기 위한 getDealer(), getPlayers() 메소드 추가

1Consumption commented:
피드백 반영하여 수정했습니다.
enum의 rawValue에 직접 접근하여 rawValue만큼 반복되는 동작을 하는 것을 enum 내에 forEach문을 사용하여 대체할 수 있다는 점을 알게 되었습니다.
또한 테스트 코드를 위해 사용하지 않는 메소드를 만들필요는 없다는 것을 알게 되었습니다.
하지만 상속 대신 프로토콜을 채택하는 경우는 잘 이해가 가지 않았습니다. 제 생각에 player나 dealer의 ownCardDeck 프로퍼티는 private 해야하는데, 프로토콜을 채택하게되면 해당 프로퍼티가 private으로 선언이 되지않아 computed property를 인터페이스로 사용하는 방식으로 구현했습니다. 이 경우 중복되는 코드도 생기고, 불필요한 프로퍼티도 생기기 때문에 상속보다 나은 점은 잘 모르겠습니다.

godrm commented:
이와중에 왜 머지 커밋이 포함되어 있나요?

godrm commented on CardGameApp/CardGameApp/model/PokerGame.swift:
```diff
@@ -84,4 +84,12 @@ class PokerGame{
+    
+    func getDealer() -> Dealer{
+        return self.dealer
+    }
```
private 으로 선언한다고 해서 다시 getter 형식으로 값을 가져가는 구조는 지양하세요.
반복해서 설명하지만 외부에서 객체 내부 속성 값을 가져가서 처리하려 하지말고, 
객체가 처리하는 메소드를 구현하려고 노력하세요.

godrm commented on CardGameApp/CardGameApp/model/PokerGame.swift:
```diff
@@ -73,10 +73,8 @@ class PokerGame{
-extension PokerGame:CustomStringConvertible{
-    var description: String {
+    
+    func printPlayersCard() -> String{
```
이 메소드가 있는 게 좋은 선택이 아닙니다. 결과를 그대로 만들어서 전달하면 다음 단계에서 출력할 수가 없습니다.

godrm commented:
머지할테니 다음 단계를 진행하면서 개선해보세요.
